### ISSUE TITLE
the result of SELECT with nested subquery is inconsistent with pg

### ISSUE THREAD

**Issue by wzzzzd** [risingwavelabs/risingwave#4355]
**Describe the bug**
```sql
create table t1 (a int, b int);
insert into t1 values (1, NULL), (2, 3), (NULL, NULL);
select a, (select count(*) from t1 where t1.a is distinct from t.b) from t1 as t order by 1;
select a, (select count(*) from t1 where t1.a <> t.b) from t1 as t order by 1;
```

In RisingWave, 
```
dev=> select a, (select count(*) from t1 where t1.a is distinct from t.b) from t1 as t order by 1;
 a | ?column? 
---+----------
 1 |         
 2 |        3
   |         
(3 rows)
dev=> select a, (select count(*) from t1 where t1.a <> t.b) from t1 as t order by 1;
 a | ?column? 
---+----------
 1 |         
 2 |        2
   |         
(3 rows)
```

In PostgreSQL,
```
postgres=# select a, (select count(*) from t1 where t1.a is distinct from t.b) from t1 as t order by 1;
 a | count 
---+-------
 1 |     2
 2 |     3
   |     2
(3 rows)
postgres=# select a, (select count(*) from t1 where t1.a <> t.b) from t1 as t order by 1;
 a | count 
---+-------
 1 |     0
 2 |     2
   |     0
(3 rows)
```

**To Reproduce**
As above.

**Expected behavior**
The outputs of RisingWave and pg should be the same.

**Additional context**
```
dev=> explain select a, (select count(*) from t1 where t1.a is distinct from t.b) from t1 as t order by 1;
                                      QUERY PLAN                                      
--------------------------------------------------------------------------------------
 BatchExchange { order: [t1.a ASC], dist: Single }
   BatchSort { order: [t1.a ASC] }
     BatchHashJoin { type: LeftOuter, predicate: t1.b = t1.b }
       BatchExchange { order: [], dist: HashShard(t1.b) }
         BatchScan { table: t1, columns: [a, b] }
       BatchHashAgg { group_key: [t1.b], aggs: [count] }
         BatchExchange { order: [], dist: HashShard(t1.b) }
           BatchNestedLoopJoin { type: Inner, predicate: IsDistinctFrom(t1.a, t1.b) }
             BatchExchange { order: [], dist: Single }
               BatchHashAgg { group_key: [t1.b], aggs: [] }
                 BatchExchange { order: [], dist: HashShard(t1.b) }
                   BatchScan { table: t1, columns: [b] }
             BatchExchange { order: [], dist: Single }
               BatchScan { table: t1, columns: [a] }
(14 rows)
```

**Comment by wzzzzd** [comment]
It seems the root cause of #4233, #4203 and this issue are the same😇
If I set `match_null` flag of `HashJoinExecutor` to true, the result becomes correct.

**Comment by chenzl25** [comment]
Fixed By #4386 and #5021.
